### PR TITLE
On cart preview, ticket style, make every product line clickable #1309

### PR DESCRIPTION
**Pull request by ithiame**

### CODE REVIEW

File: src/routes/(app)/cart/[id]/+page.server.ts [hunk 1]
```diff
@@ -112,6 +112,40 @@ export const actions = {
...
+				$set: { updatedAt: new Date() }
+			});
+			throw error(404, 'This product does not exist');
+		}
+		if (!product) {
+			await collections.carts.updateOne(userQuery(userIdentifier(locals)), {
+				$pull: { items: { productId: params.id } },
+				$set: { updatedAt: new Date() }
+			});
+			throw error(404, 'This product does not exist');
```
Comment (coyotte508) [inline src/routes/(app)/cart/[id]/+page.server.ts:131]: Duplicate code with just above

File: src/routes/(app)/pos/touch/+page.svelte [hunk 2]
```diff
@@ -45,15 +48,45 @@
...
+							};
+						}}
+					>
+						<input type="hidden" name="note" value={item.note?.note || ''} />
+						<input type="hidden" name="quantity" value={item.quantity} />
+						<button type="submit" class="text-2xl" on:click={() => addNoteToItem(i)}>
+							{item.quantity} X {item.product.name.toUpperCase()}
+						</button><br />
+						{item.note?.note ? '+' + item.note?.note : ''}
+					</form>
```
Comment (coyotte508) [inline src/routes/(app)/pos/touch/+page.svelte:89]: This form should just be the note, `quantity` shouldn't be involved

File: src/routes/(app)/pos/touch/+page.svelte [hunk 3]
```diff
@@ -45,15 +48,45 @@
...
 	$: currentPage = Math.floor(next / POS_PRODUCT_PAGINATION) + 1;
+
+	function addNoteToItem(index: number) {
+		const notePrompt = prompt('enter a comment:');
+		if (notePrompt) {
+			items = items.map((item, i) =>
+				i === index ? { ...item, note: { note: notePrompt, internal: true } } : item
+			);
+		}
+	}
```
Comment (coyotte508) [inline src/routes/(app)/pos/touch/+page.svelte:59]: You should pass `item` directly

and do:

```ts
item.note = ...;
items = [...items];
```

File: src/routes/(app)/cart/[id]/+page.server.ts [hunk 4]
```diff
@@ -112,6 +112,33 @@ export const actions = {
...
+			})
+			.parse({
+				note: formData.get('note')
+			});
+
+		await addToCartInDb(product, quantity, {
+			user: userIdentifier(locals),
+			totalQuantity: true,
+			note: note
+		});
```
Comment (coyotte508) [inline src/routes/(app)/cart/[id]/+page.server.ts:140]: I don't think you should call `addToCartInDB`

Instead you can do:

```ts
const cart = await collections.cart.findOne(userQuery(userIdentifier(locals));

if (!cart) {
  throw ...;
}

const res = await collections.cart.updateOne({_id: cart._id, "items.productId": params.id}, {
	$set: {"$product.internalNote": ...}
});

if (!res.matchedCount) {
  throw ... // cart in DB doesn't have the product
}
```

You can remove all the product checking code.

File: src/routes/(app)/cart/[id]/+page.server.ts [hunk 5]
```diff
@@ -112,6 +112,33 @@ export const actions = {
...
+				$pull: { items: { productId: params.id } },
+				$set: { updatedAt: new Date() }
+			});
+			throw error(404, 'This product does not exist');
+		}
+		const formData = await request.formData();
+		const quantity = 1;
+		const { note } = z
+			.object({
+				note: z.string()
```
Comment (coyotte508) [inline src/routes/(app)/cart/[id]/+page.server.ts:130]:
```suggestion
				note: z.string().trim().min(1)
```

File: src/lib/types/Cart.ts [hunk 6]
```diff
@@ -24,6 +24,7 @@ export interface Cart extends Timestamps {
 		customPrice?: { amount: number; currency: Currency };
 		reservedUntil?: Date;
 		depositPercentage?: number;
+		note?: { note: string; internal: boolean };
```
Comment (coyotte508) [inline src/lib/types/Cart.ts:27]: Let's use `internalNote` to store the note. That way later if we have a user note, it won't conflict

```suggestion
		internalNote?: { value: string; updatedAt: Date, updatedById: User["_id"] };
```

**Review by coyotte508 — COMMENTED** [review]
Let's use `internalNote` to store the note.

File: src/routes/(app)/cart/[id]/+page.server.ts [hunk 7]
```diff
@@ -113,5 +113,37 @@ export const actions = {
 		});
 
 		throw redirect(303, request.headers.get('referer') || '/cart');
+	},
+	addNote: async ({ locals, params, request }) => {
+		const cart = await collections.carts.findOne(userQuery(userIdentifier(locals)));
+
+		if (!cart) {
+			throw error(404, 'This product is not in the cart');
```
Comment (coyotte508) [inline src/routes/(app)/cart/[id]/+page.server.ts:121]:
```suggestion
			throw error(404, 'No cart found for user');
```

**Review by coyotte508 — APPROVED** [review]

**Comment by Tirodem** [comment]
"If a comment already exist, use the same form and pre-fill with the current value."
Even if there's a beginning :
image
We need the new component to display the pop-in and the possibility to edit the current value of the line comment.

**Comment by Tirodem** [comment]
+ clic zone should not be here :
image
If should be all over there :
image

**Comment by Tirodem** [comment]
Cancelling a comment creation triggers this error ❌ 
image

**Pull request merged**